src/NumericalAlgorithms/SphericalHarmonics/StrahlkorperFunctions.cpp

```diff
@@ -467,17 +468,52 @@ void time_deriv_of_strahlkorper(
         previous_strahlkorpers) {
   std::deque<double> times{};
   std::deque<const DataVector*> coefficients{};
+  std::deque<DataVector> prolonged_or_restricted_coefficients{};
 
-  // Can't take time deriv of 1 strahlkorper
+  // Can't take time deriv of 1 strahlkorper, so just zero the time derivative's
+  // coefficients
   if (previous_strahlkorpers.size() == 1) {
-    time_deriv->coefficients() = DataVector{
-        previous_strahlkorpers.front().second.coefficients().size(), 0.0};
+    time_deriv->coefficients() = 0.0;
     return;
   }
 
-  for (const auto& [time, strahlkorper] : previous_strahlkorpers) {
-    times.emplace_back(time);
-    coefficients.emplace_back(&strahlkorper.coefficients());
+  // Find the min and max l_max of each previous Strahlkorper, along with the
+  // index of max(l_max). Getting both min(l_max) and max(l_max) enables
+  // skipping the prolong/restrict call when min(l_max) == max(l_max), and
+  // getting the index of max(l_max) enables skipping the prolong/restrict
+  // call on whichever previous Strahlkorper has the highest l_max.
+  const auto [min_l_it, max_l_it] = std::minmax_element(
+      previous_strahlkorpers.begin(), previous_strahlkorpers.end(),
+      [](const auto& a, const auto& b) {
+        return a.second.l_max() < b.second.l_max();
+      });
+  const size_t max_prev_l = max_l_it->second.l_max();
+  const size_t min_prev_l = min_l_it->second.l_max();
+  // OK to static_cast to size_t from long here because
+  // max_l_it is inside the std::deque container, so distance
+  // from previous_strahlkorpers.begin() to max_l_it must be nonnegative.
+  const auto max_prev_l_index = static_cast<size_t>(
+      std::distance(previous_strahlkorpers.begin(), max_l_it));
```

Comment:
Okay, makes sense to me now, thanks for cleaning that up!

I think now you can do:
```cpp
const auto max_l = *std::max(revious_strahlkorpers.begin(), previous_strahlkorpers.end(),
      [](const auto& a, const auto& b) {
        return a.second.l_max() < b.second.l_max();
      });
```
and then remove the `min_prev_l != max_prev_l and` below because the check should only be true if `previous_strahlkorpers[i].second.l_max() != max_prev_l` regardless of what the min/max are.

I would also add the following to the loop below:
```cpp
ASSERT(previous_strahlkorpers[i].second.l_max() <= max_prev_l, "Calculated incorrect max ell, " << max_prev_l << " but got " << previous_strahlkorpers[i].second.l_max());
```

Reply:
Thanks for the suggestion! I agree this is better, and wish I'd seen that myself when doing the cleanup. But I still ned `std::max_element`, not just `std::max`, because when I prolong or restrict, I need to know which index of the previous surfaces to use to get the `ylm_spherepack` that I am prolonging to.

Also, I see no way to trigger the assert from outside the function, so I didn't modify my unit test. But if I missed something and there is a way to test it in the unit test, just let me know!

Reply:
Oh right, sorry I missed that you need the index. This looks great!

I don't think there is a way to test the ASSERT from outside the function, it's just meant to be a sanity check that you calculated it correctly :)